File: que78.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> subset = new ArrayList<>();

        createSubset(nums, 0, result, subset);
        return result;
    }
    private void createSubset(int[] nums, int idx, List<List<Integer>> result, List<Integer> subset){
        if(idx == nums.length){
            result.add(new ArrayList<>(subset));
            return;
        }
        subset.add(nums[idx]);
        createSubset(nums, idx+1, result, subset);

        subset.remove(subset.size()-1);
        createSubset(nums, idx+1, result, subset);
    }
}

File: que78.cpp
class Solution {
public:
    void getAllSubsets(vector<int>& nums,vector<int>& ans,int i,vector<vector<int>> &allSubsets){
        //base case
        if(i==nums.size()){
            allSubsets.push_back({ans});
            return;
        }
        
        //inclusion call
        ans.push_back(nums[i]);
        getAllSubsets(nums,ans,i+1,allSubsets);

        //backtrack
        ans.pop_back();
        
        //exclusion call
        getAllSubsets(nums,ans,i+1,allSubsets);
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> ans;
        vector<vector<int>> allSubsets;
        getAllSubsets(nums,ans,0,allSubsets);
        return allSubsets;
    }
};
